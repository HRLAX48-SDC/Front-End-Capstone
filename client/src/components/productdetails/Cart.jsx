import React, { useState, useEffect } from 'react';
import Favorite from './Favorite.jsx';
import AddToCart from './AddToCart.jsx';
import axios from 'axios';
import API_KEY from '../../config/config.js';

function Cart(props) {
  const [size, setSize] = useState('SELECT SIZE');
  const [quantity, setQuantity] = useState(0);
  const [notice, setNotice] = useState(false);

  useEffect(() => {
    reset();
  }, [props]);

  const reset = () => {
    setSize('SELECT SIZE');
    setQuantity(0);
  };

  const selectSize = (e) => {
    setSize(e.target.value);
    setNotice(false);
  };

  const selectQuantity = () => {
    if (props.style.length && size !== 'SELECT SIZE') {
      var quantity = Object.values(props.style[props.styleNum].skus).filter(
        (item) => {
          return item.size === size;
        }
      )[0].quantity;
      if (quantity > 15) {
        quantity = 15;
      }
      var numList = Array.from({ length: quantity }, (_, i) => i + 1);
      return numList;
    }
  };

  const pickQuantity = (e) => {
    setQuantity(e.target.value);
  };

  const sku = () => {
    var skus = props.style[props.styleNum].skus;
    var skus_id;
    for (var key in skus) {
      if (skus[key].size === size) {
        skus_id = key;
      }
    }
    return skus_id;
  };

  const addToCart = () => {
    var data = {
      sku_id: sku(),
      count: quantity,
    };
    axios
      .post('https://13.57.182.185/api/cart', data)
      .then((response) => {
        alert('Added to Cart!');
      })
      .catch((err) => {
        console.error(err);
      });
  };

  const handleClick = () => {
    if (size === 'SELECT SIZE') {
      setNotice(true);
    }
    if (size !== 'SELECT SIZE' && quantity !== 0) {
      addToCart();
      reset();
    }
  };

  const checkQuantity = () => {
    const sizeAndQuantity = Object.values(props.style[props.styleNum].skus);
    if (sizeAndQuantity[0].size) {
      return (
        <select onChange={selectSize} className='p_select'>
          <option>SELECT SIZE</option>
          {sizeAndQuantity.map((item, index) => (
            <option key={index} value={item.size} className='p_dropdown'>
              {item.size}
            </option>
          ))}
        </select>
      );
    } else {
      return (
        <select className='p_select'>
          <option>OUT OF STOCK</option>
        </select>
      );
    }
  };

  return (
    <div>
      {notice && <div className='p_notice'>Please select size</div>}
      <div className='p_flex_container'>
        <div className='p_flex_child_size'>
          {props.style.length && checkQuantity()}
        </div>

        <div className='p_flex_child_quantity'>
          {size === 'SELECT SIZE' ? (
            <select className='p_select'>
              <option>-</option>
            </select>
          ) : (
            <select onChange={pickQuantity} className='p_select'>
              {selectQuantity().map((num) => (
                <option key={num} value={num}>
                  {num}
                </option>
              ))}
            </select>
          )}
        </div>
      </div>

      {props.style.length &&
      Object.values(props.style[props.styleNum].skus)[0].size ? (
        <AddToCart handleClick={handleClick} />
      ) : null}
      <Favorite style={props.style} />
    </div>
  );
}

export default Cart;
