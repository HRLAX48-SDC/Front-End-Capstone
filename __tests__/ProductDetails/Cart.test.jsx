import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { configure, mount, shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
import Cart from '../../client/src/components/productdetails/Cart.jsx';

test('renders Cart component correctly', () => {
  const style = [
    {
      style_id: 253620,
      name: "Forest Green & Black",
      original_price: "140.00",
      skus: {
        1471554: {quantity: 8, size: 'XS'},
        1471555: {quantity: 16, size: 'S'},
        1471556: {quantity: 17, size: 'M'},
        1471557: {quantity: 10, size: 'L'},
        1471558: {quantity: 15, size: 'XL'},
        1471559: {quantity: 4, size: 'XL'}
      }
    }
  ];
  const styleNum = 0;
  const cart = renderer.create(<Cart style={style} styleNum={styleNum} />).toJSON();
  expect(cart).toMatchSnapshot();
});

test('Renders the status component', ()=> {
  const style = [
    {
      style_id: 253620,
      name: "Forest Green & Black",
      original_price: "140.00",
      skus: {
        1471554: {quantity: 8, size: 'XS'},
        1471555: {quantity: 16, size: 'S'},
        1471556: {quantity: 17, size: 'M'},
        1471557: {quantity: 10, size: 'L'},
        1471558: {quantity: 15, size: 'XL'},
        1471559: {quantity: 4, size: 'XL'}
      }
    }
  ];
  const styleNum = 0;
  const wrapper = shallow(<Cart styleNum={styleNum} style={style} />);
  expect(wrapper).toContainMatchingElements(2,'select');
  expect(wrapper).toContainMatchingElements(8,'option');
})