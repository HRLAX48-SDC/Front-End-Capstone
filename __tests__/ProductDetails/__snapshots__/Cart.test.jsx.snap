// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Cart component correctly 1`] = `
<div>
  <div
    className="p_flex_container"
  >
    <div
      className="p_flex_child_size"
    >
      <select
        className="p_select"
        onChange={[Function]}
      >
        <option>
          SELECT SIZE
        </option>
        <option
          className="p_dropdown"
          value="XS"
        >
          XS
        </option>
        <option
          className="p_dropdown"
          value="S"
        >
          S
        </option>
        <option
          className="p_dropdown"
          value="M"
        >
          M
        </option>
        <option
          className="p_dropdown"
          value="L"
        >
          L
        </option>
        <option
          className="p_dropdown"
          value="XL"
        >
          XL
        </option>
        <option
          className="p_dropdown"
          value="XL"
        >
          XL
        </option>
      </select>
    </div>
    <div
      className="p_flex_child_quantity"
    >
      <select
        className="p_select"
      >
        <option>
          -
        </option>
      </select>
    </div>
  </div>
  <button
    className="btn btn-default btn-sm p_flex_child_cart btn btn-dark"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <b>
       Add to Cart +
    </b>
  </button>
  <button
    className="btn btn-default btn-md p_fav btn btn-light"
    disabled={false}
    id="p_like_button"
    onClick={[Function]}
    type="button"
  >
    <i
      className="far fa-star"
    />
  </button>
</div>
`;
